src/main/query/where.c

```diff
@@ -63,352 +63,265 @@ static int AerospikeQuery_Where_Add(AerospikeQuery *self, PyObject *py_ctx,
+                                    index_type, in_datatype, val1_int,
+                                    val2_int);
         }
         else {
-            // If it ain't right, raise and error
-            as_error_update(&err, AEROSPIKE_ERR_PARAM,
-                            "range predicate type not supported");
-            PyObject *py_err = NULL;
-            error_to_pyobject(&err, &py_err);
-            PyErr_SetObject(PyExc_Exception, py_err);
-            rc = 1;
-            break;
+            as_query_where_with_ctx(&self->query, bin, pctx, predicate,
+                                    index_type, in_datatype, val1);
+        }
+
+        if (in_datatype == AS_INDEX_STRING ||
+            in_datatype == AS_INDEX_GEO2DSPHERE) {
+            self->query.where.entries[0].value.string_val._free = true;
         }
-        break;
     }
-    default: {
+    else {
         // If it ain't supported, raise and error
         as_error_update(&err, AEROSPIKE_ERR_PARAM, "unknown predicate type");
         PyObject *py_err = NULL;
         error_to_pyobject(&err, &py_err);
         PyErr_SetObject(PyExc_Exception, py_err);
-        rc = 1;
-        break;
+        goto CLEANUP_ON_ERROR2;
     }
+
+    if (ctx_in_use) {
+        self->query.where.entries[0].ctx_free = true;
     }
 
-    if (rc) {
-        assert(false);
-        if (ctx_in_use) {
-            as_cdt_ctx_destroy(pctx);
-        }
-        if (pctx) {
-            cf_free(pctx);
-        }
+    return 0;
+
+CLEANUP_ON_ERROR2:
+
+    // The values end up not being used by as_query
+    if (val1_str) {
+        free(val1_str);
     }
-    else if (ctx_in_use) {
-        self->query.where.entries[0].ctx_free = true;
+    if (val1_bytes) {
+        free(val1_bytes);
+    }
+
+CLEANUP_ON_ERROR1:
+
+    // The ctx ends up not being used by as_query
+    if (ctx_in_use) {
+        as_cdt_ctx_destroy(pctx);
+    }
+    if (pctx) {
+        cf_free(pctx);
     }
-    return rc;
+
+    return 1;
 }
 
+#define PREDICATE_INVALID_ERROR_MSG1 "predicate is invalid."
+#define PREDICATE_INVALID_ERROR_MSG2 "Failed to fetch predicate information"
+
 AerospikeQuery *AerospikeQuery_Where_Invoke(AerospikeQuery *self,
-                                            PyObject *py_arg1,
-                                            PyObject *py_arg2)
+                                            PyObject *py_ctx,
+                                            PyObject *py_predicate)
 {
-    as_error err;
-    int rc = 0;
 
+    as_error err;
     as_error_init(&err);
 
-    if (PyTuple_Check(py_arg2) && PyTuple_Size(py_arg2) > 1 &&
-        PyTuple_Size(py_arg2) <= 6) {
+    // Parse predicate tuple
 
-        Py_ssize_t size = PyTuple_Size(py_arg2);
+    if (!PyTuple_Check(py_predicate)) {
+        as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                        PREDICATE_INVALID_ERROR_MSG1);
+        goto CLEANUP;
+    }
+    Py_ssize_t size = PyTuple_Size(py_predicate);
```

Comment:
```suggestion
    Py_ssize_t predicate_size = PyTuple_Size(py_predicate);
```
nit: I think this would make understanding the code easier.

Reply:
Agreed, done.